lib/tlsutil/tlsutil.go

```diff
@@ -117,42 +117,52 @@ func NewCertificate(certFile, keyFile, commonName string, lifetimeDays int) (tls
 
 	derBytes, err := x509.CreateCertificate(rand.Reader, &template, &template, publicKey(priv), priv)
 	if err != nil {
-		return tls.Certificate{}, errors.Wrap(err, "create cert")
+		return nil, nil, errors.Wrap(err, "create cert")
 	}
 
-	certOut, err := os.Create(certFile)
+	certBlock := &pem.Block{Type: "CERTIFICATE", Bytes: derBytes}
+	keyBlock, err := pemBlockForKey(priv)
 	if err != nil {
-		return tls.Certificate{}, errors.Wrap(err, "save cert")
-	}
-	err = pem.Encode(certOut, &pem.Block{Type: "CERTIFICATE", Bytes: derBytes})
-	if err != nil {
-		return tls.Certificate{}, errors.Wrap(err, "save cert")
-	}
-	err = certOut.Close()
-	if err != nil {
-		return tls.Certificate{}, errors.Wrap(err, "save cert")
+		return nil, nil, errors.Wrap(err, "save key")
 	}
 
-	keyOut, err := os.OpenFile(keyFile, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
-	if err != nil {
-		return tls.Certificate{}, errors.Wrap(err, "save key")
-	}
+	return certBlock, keyBlock, nil
+}
 
-	block, err := pemBlockForKey(priv)
+// NewCertificate generates and returns a new TLS certificate, saved to the given PEM files.  If any file name is empty, the contents are kept only in memory.
+func NewCertificate(certFile, keyFile string, commonName string, lifetimeDays int) (tls.Certificate, error) {
+	certBlock, keyBlock, err := generateCertificate(commonName, lifetimeDays)
 	if err != nil {
-		return tls.Certificate{}, errors.Wrap(err, "save key")
+		return tls.Certificate{}, err
 	}
 
-	err = pem.Encode(keyOut, block)
-	if err != nil {
-		return tls.Certificate{}, errors.Wrap(err, "save key")
+	if certFile != "" {
+		certOut, err := os.Create(certFile)
+		if err != nil {
+			return tls.Certificate{}, errors.Wrap(err, "save cert")
+		}
+		if err = pem.Encode(certOut, certBlock); err != nil {
+			return tls.Certificate{}, errors.Wrap(err, "save cert")
+		}
+		if err = certOut.Close(); err != nil {
+			return tls.Certificate{}, errors.Wrap(err, "save cert")
+		}
 	}
-	err = keyOut.Close()
-	if err != nil {
-		return tls.Certificate{}, errors.Wrap(err, "save key")
+
+	if keyFile != "" {
+		keyOut, err := os.OpenFile(keyFile, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
+		if err != nil {
+			return tls.Certificate{}, errors.Wrap(err, "save key")
+		}
+		if err = pem.Encode(keyOut, keyBlock); err != nil {
+			return tls.Certificate{}, errors.Wrap(err, "save key")
+		}
+		if err = keyOut.Close(); err != nil {
+			return tls.Certificate{}, errors.Wrap(err, "save key")
+		}
 	}
 
-	return tls.LoadX509KeyPair(certFile, keyFile)
+	return tls.X509KeyPair(pem.EncodeToMemory(certBlock), pem.EncodeToMemory(keyBlock))
```

Comment:
I dislike the pass empty string to not save. Why can't it just be two methods? I guess you can't go from tls.Cert back to pem blocks?

Reply:
Yes, at least it wasn't obvious to me how to do that. Something happens with certificate chains and parsing and you cannot recover the original form. If you find a way, I'm all ears.

Reply:
It's about the call signature, where both empty strings and two methods isn't particularly nice, but I too would prefer the latter. Something like e.g.

```
func NewCertificate(certFile, keyFile string, commonName string, lifetimeDays int) (tls.Certificate, error)

func NewCertificateMemoryOnly(commonName string, lifetimeDays int) (tls.Certificate, error)
```

Or alternatively export `tlsutil.generateCertificate`, and have

```
func NewCertificate(certBlock, keyBlock *pem.Block) (tls.Certificate, error)

func SaveCertificate(certBlock, keyBlock *pem.Block) error
```
or possibly encapsulate the two `*pem.Block` variables in a type in tlsutil to not even expose that "implementation detail".

Reply:
I find it strange if a caller needs to call `GenerateCertificate()` and then `NewCertificate()` to actually use it.  Especially when the `tlsutil` package is obviously meant to make the process easier.

So I went with a separate function `NewCertificateInMemory()`. OK?